test: Performance benchmark

davidkornel opened the pull request:
An example script to evaluate the performance of STUNner in and outside of Kubernetes.

levaitamas commented on examples/benchmark/README.md:
```diff
@@ -0,0 +1,91 @@
+
+## Installation
+
+Install it in case you would like to benchmark your Kubernetes setup. If you want to benchmark locally skip this step.
```
I would add 2 subsections to Installation:

```
## Install locally

You are good to go. No installation steps required.

## Install on Kubernetes

....
```

levaitamas commented on examples/benchmark/README.md:
```diff
@@ -0,0 +1,91 @@
+
+Install it in case you would like to benchmark your Kubernetes setup. If you want to benchmark locally skip this step.
+
+Install the STUNner gateway operator and STUNner ([more info](https://github.com/l7mp/stunner-helm)):
```
Isn't this step coverd by the pre-requirements?

davidkornel replied:
You're right. I removed the relevant part from the Pre-requirements. I wanted to keep the install guide to make this whole thing easier for the users to understand.

levaitamas commented on examples/benchmark/README.md:
````diff
@@ -0,0 +1,91 @@
+kubectl apply -f performance-stunner.yaml
+```
+
+## Usage
````
I would rename this section to `Executing measurements`

levaitamas commented on examples/benchmark/README.md:
```diff
@@ -0,0 +1,91 @@
+
+## Usage
+
+### Parameters
```
`Helper script parameters`

and please introduce the script: "We bundle a helper script for executing performance measurements. ..."

levaitamas commented on examples/benchmark/README.md:
````diff
@@ -0,0 +1,91 @@
+```
+./benchmark.sh 5 5 1000 100000000 k8s
+```
+## Advice
````
I would copy the Help section from other READMEs in case anyone gets lost. :)

levaitamas commented on examples/benchmark/benchmark.sh:
```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
+
```
perhaps add the MIT license blob?

levaitamas commented on examples/benchmark/benchmark.sh:
```diff
@@ -0,0 +1,61 @@
+packet_size=$3
+bandwidth=$4
+platform=$5
+
```
Please implementI  `-h` / `--help` for the tl;dr folks.

davidkornel replied:
Thanks for the idea, your suggestion made me change the positional parameters to flags and set default values. This makes the whole script more convenient to use IMO.

levaitamas commented on examples/benchmark/README.md:
```diff
@@ -0,0 +1,91 @@
+
+The below command will open:
+* a `stunnerd` UDP listener at `127.0.0.1:5001`
+* one or more `turncat` clients at `127.0.0.1:90XY` to open a connection through STUNner to the iperf server
```
I would add a comment (90XY are ports used for measurement purposes starting from 9000)

levaitamas commented on examples/benchmark/benchmark.sh:
```diff
@@ -0,0 +1,61 @@
+sleep $(($eval_time+2))
+killall main -w &> /dev/null
+killall iperf -w &> /dev/null
+exit
```
I would recommend running the code through a linter (e.g., https://www.shellcheck.net/ )

levaitamas commented on examples/benchmark/iperf-server.yaml:
```diff
@@ -0,0 +1,39 @@
+        app: iperf-server
+    spec:
+      containers:
+      - image: l7mp/net-debug:0.5.4
```
perhaps use the `latest`?

levaitamas commented on examples/benchmark/README.md:
```diff
@@ -0,0 +1,91 @@
+With the help of this guide you are able to take performance measurments in your setup using STUNner. Both running STUNner locally (outside of Kubernetes) and running STUNner in Kubernetes can be evaluated.
+
+Compare the locally measured results to the result measured in Kubernetes and figure out the overhead cost. The extra cost of your cluster's networking may surprise you in terms of extra delay or more packet drops using the same bandwidth.
+
```
If you could give a rough time estimate (for install and measure) that would be great extra.

levaitamas reviewed (commented):
Nice! I added a couple of minor comments.

nmate commented:
Probe

davidkornel commented:
@nmate @levaitamas Could you give another review?

levaitamas commented:
@davidkornel thanks for addressing the comments, and the figures are great! I still have some ideas/comments. Do you mind if I do some edits on your code/branch?

davidkornel commented:
> @davidkornel thanks for addressing the comments, and the figures are great! I still have some ideas/comments. Do you mind if I do some edits on your code/branch?

I don't mind at all, if I can learn new stuff that's always good. Go ahead. Curious to see your ideas.

davidkornel commented on examples/benchmark/README.md:
```diff
@@ -139,8 +144,7 @@ Results
 
-In this case the output will contain one or more summarized tests. If the script were rerun twice, two `iperf` outputs would be combined. 
-You should see something like this:
+In case of a Kubernetes measurement, the output contains one or more summarized tests. If the script was rerun twice, two `iperf` outputs would be combined. You should an output similar to this:
```
Missing verb in the last sentence. Perhaps, "you should see an output..."

davidkornel replied:
fixed

nmate commented on examples/benchmark/benchmark.sh:
```diff
@@ -0,0 +1,113 @@
+#!/bin/bash
```
I would also consider to add "-e" flag to stop the execution in case of an error

nmate commented on examples/benchmark/benchmark.sh:
```diff
@@ -0,0 +1,113 @@
+
+if [[ $platform == "local" ]]; then
+    STUNNER_PUBLIC_ADDR="127.0.0.1"
+    STUNNER_PUBLIC_PORT="5001"
```
port 5001 hopefully will be free

nmate commented on examples/benchmark/benchmark.sh:
```diff
@@ -0,0 +1,113 @@
+    sleep 2
+
+elif [[ $platform == "k8s" ]]; then
+    STUNNER_PUBLIC_ADDR=$(kubectl get cm stunnerd-config -n stunner -o jsonpath='{.data.stunnerd\.conf}' | jq -r .listeners[0].public_address)
```
Is the format the same in case of the standalone deployed Stunner ConfigMap and the operator based version? If not, you shall mention in the README which version is supported by the script.

nmate reviewed (commented):
Minor comments, otherwise I like it!

nmate commented on examples/benchmark/README.md:
```diff
@@ -0,0 +1,190 @@
+Platform: k8s
+Results
+[ ID] Interval            Transfer     Bandwidth        Jitter   Lost/Total  Latency avg/min/max/stdev PPS  inP NetPwr
+[  3] 0.0000-4.9589 sec   613 KBytes  1.01 Mbits/sec   0.756 ms    0/  628 (0%)  7.780/ 7.094/56.286/ 4.166 ms  127 pps  985 Byte 16.28
```
Now I see its ms and the meaning of columns. My bad, resolved.
